source/opt/fold_spec_const_op_composite.cpp

```diff
@@ -0,0 +1,565 @@
+            id_to_const_val_[iter->second->AsComposite()->nested[d]]
+                ->AsScalar()
+                ->word);
+      }
+      uint32_t one_dimension_result =
+          Operate(static_cast<SpvOp>(inst->GetSingleWordInOperand(0)),
+                  one_dimension_operand_values_in_raw_words);
+      auto one_dimension_result_val =
+          std::unique_ptr<Value>(new ScalarValue(one_dimension_result));
+      uint32_t new_id = GetNewId();
+      id_to_const_val_[new_id] = std::move(one_dimension_result_val);
+      result->AsComposite()->nested.push_back(new_id);
+      new_component_ids.emplace_back(new_id);
+    }
+    return std::make_tuple(std::move(result), new_component_ids);
+  }
+}
+
+ir::Module::inst_iterator FoldSpecConstantOpAndCompositePass::InsertConstantAt(
+    const analysis::Type* type, uint32_t result_id,
+    ir::Module::inst_iterator inst_iter) {
+  std::unique_ptr<ir::Instruction> new_inst;
+  uint32_t type_id = type_mgr_->GetId(type);
+  auto iter = id_to_const_val_.find(result_id);
+  assert(iter != id_to_const_val_.end() &&
+         "The value for the new constant should be added to id_to_const_val_ "
+         "map before creating and inserting its instruction");
+
+  if (iter->second->AsScalar()) {
+    uint32_t word = iter->second->AsScalar()->word;
+    if (type->AsBool()) {
+      new_inst.reset(new ir::Instruction(word != static_cast<uint32_t>(false)
+                                             ? SpvOp::SpvOpConstantTrue
+                                             : SpvOp::SpvOpConstantFalse,
+                                         type_id, result_id, {}));
+    } else {
+      ir::Operand literal(spv_operand_type_t::SPV_OPERAND_TYPE_LITERAL_INTEGER,
+                          {word});
+      new_inst.reset(new ir::Instruction(SpvOp::SpvOpConstant, type_id,
+                                         result_id, {literal}));
+    }
+  } else {
+    // Composite type.
+    CompositeValue* new_val = iter->second->AsComposite();
+    std::vector<ir::Operand> operands;
+    for (uint32_t component_id : new_val->nested) {
+      operands.emplace_back(SPV_OPERAND_TYPE_ID,
+                            std::initializer_list<uint32_t>{component_id});
+    }
+    new_inst.reset(new ir::Instruction(SpvOp::SpvOpConstantComposite, type_id,
+                                       result_id, std::move(operands)));
+  }
+
+  // update def use info
+  def_use_mgr_->AnalyzeInstDefUse(new_inst.get());
+  return inst_iter.InsertBefore(std::move(new_inst));
+}
+
+std::unique_ptr<Value> FoldSpecConstantOpAndCompositePass::GetConstValue(
+    ir::Instruction* inst) {
+  switch (inst->opcode()) {
+    case SpvOp::SpvOpConstant: {
+      assert(inst->NumInOperands() == 1 &&
+             "OpConstant instruction should have only one in-operand");
+      analysis::Type* ty = GetType(inst);
+      // Only support 32 bit integer/float scalar.
+      if (ty->AsInteger() && ty->AsInteger()->width() != 32) {
```

Comment:
Sorry, my previous comment is too ambiguous. I think this will allow non integer/fp types? I would expect something similar to

```
if (ty->AsInteger()) {
  if (...->width() != 32) return nullptr;
} else {
  return nullptr;
}